k_thread_abort/join() rewrite & attendant fixes

andyross opened the pull request:
Comparatively big series that reworks abort and join to fix some fundamental races on SMP (see #32344).   The new version is (heh) hopefully correct, but should absolutely be easier to review and reason about.  It definitely doesn't share the same misfeatures we saw earlier.

The early patches in the series are fixes to related bugs I stumbled on while working this mixed with refactoring needed for later changes.  Note that the abort rework itself is split into two patches for review, even though that technically breaks bisectability (there's a big warning to this effect).

Commit:
kernel: Remove abort and join implementation (UNBISECTABLE)

THIS COMMIT DELIBERATELY BREAKS BISECTABILITY FOR EASE OF REVIEW.
SKIP IF YOU LAND HERE.

Remove the existing implementatoin of k_thread_abort(),
k_thread_join(), and the attendant facilities in the thread subsystem
and idle thread that support them.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Fix race between thread wakeup timeout and abort

Aborted threads will cancel their timeouts, but the timeout subsystem
isn't protected under the same lock so it's possible for a timeout to
fire just as a thread is being aborted and wake it up unexpectedly.
Check the state before blowing anything up.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Adjust locking in z_swap()

Swap was originally written to use the scheduler lock just to select a
new thread, but it would be nice to be able to rely on scheduler
atomicity later in the process (in particular it would be nice if the
assignment to cpu.current could be seen atomically).  Rework the code
a bit so that swap takes the lock itself and holds it until just
before the call to arch_switch().

Note that the local interrupt mask has always been required to be held
across the swap, so extending the lock here has no effect on latency
at all on uniprocessor setups, and even on SMP only affects average
latency and not worst case.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/thread: Initialize pended_on field of struct thread_base

This got missed, leaving garbage there for restarted threads to trip
on.  Actually I see multiple uninitialized fields, which seems odd.
This code deserves some rework, thread initialization isn't a
performance path and we should probably be zeroing the struct out.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Make z_ready_thread() safe vs. already-running threads

This is part of the scheduler API, and was always just a synchronized
wrapper around the internal ready_thread() function.  But where the
internal users seem to be careful not to call it on threads that are
not known to be already queued or running, the general users in the
IPC code seem to be less strict.

Add a simple test to detect the case where a thread is already
running.  Right now this just loops over the array of CPUs, so is O(N)
in the CPU count even though N is never more than four for us
currently.  But this is possible without modifying data structures.  A
more scalable way to do this if we ever need to run on very parallel
systems would be to use another state bit for RUNNING, or to keep a
backpointer in the thread struct to the CPU it's running on, etc...

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Add new k_thread_abort()/k_thread_join()

Add a newer, much smaller and simpler implementation of abort and
join.  No need to involve the idle thread.  No need for a special code
path for self-abort.  Joining a thread and waiting for an aborting one
to terminate elsewhere share an implementation.  All work in both
calls happens under a single locked path with no unexpected
synchronization points.

This fixes a bug with the current implementation where the action of
z_sched_single_abort() was nonatomic, releasing the lock internally at
a point where the thread to be aborted could self-abort and confuse
the state such that it failed to abort at all.

Note that the arm32 and native_posix architectures, which have their
own thread abort implementations, now see a much simplified
"z_thread_abort()" internal API.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Remove now-spurious SWAP_NONATOMIC workaround

Recent work to normalize use of the thread QUEUED state bit means that
we never attempt to remove unqueued threads from the low-level run
queue.  So the old workaround for SWAP_NONATOMIC that was trying to
detect this condition isn't necessary anymore.

Which is serendipitous, because it was written to encode some very
specific logic about the circumstances where _current could be
dequeued that I'd like to be able to break.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Clamp minimum timeslice when TICKLESS

When the kernel is TICKLESS, timeouts are set as needed, and drivers
all have some minimum amount of time before which they can reliably
schedule an interrupt.  When this happens, drivers will kick the
requested interrupt out by one tick.  This means that it's not
reliably possible to get a timeout set for "one tick in the
future"[1].

And attempting to do that is dangerous anyway.  If the driver will
delay a one-tick interrupt, then code that repeatedly tries to
schedule an imminent interrupt may end up in a state where it is
constantly pushing the interrupt out into the future, and timer
interrupts stop arriving!  The timeout layer actually has protection
against this case.

Finally getting to the point: in recent changes, the timeslice layer
lost its integration with the "imminent" test in the timeout code, so
it's now able to run into this situation: very rapidly context
switching code (or rapidly arriving interrupts) will have the effect
of infinitely[2] delaying timeouts and stalling the whole timeout
subsystem.

Don't try to be fancy.  Just clamp timeslice duration such that a
slice is 2 ticks at minimum and we'll never hit the problem.  Adjust
the two tests that were explicitly requesting very short slice rates.

[1] Of course, the tradeoff is that the tick rate can be 100x higher
or more, so on balance tickless is a huge win.

[2] Actually it only lasts until a 31 bit signed rollover in the HPET
cycle count in practice.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
And this isn't **quite** ready yet.  While it fixes the spurious abort problems, the modified timing has tripped over a different issue with timeslicing.  I hope to have a fix for that added to this series within a day; I see the root cause at least.

andyross commented on kernel/sched.c:
```diff
@@ -1426,6 +1436,124 @@ int k_thread_cpu_mask_disable(k_tid_t thread, int cpu)
+{
+	z_thread_abort(thread);
+}
+#else
```
Just noticed this empty #else, which got leftover when I removed the stuff here in favor of modifying the arch layers.  Will fix.

andyross commented:
OK, think I have it now (see the final timeslice fix which is the third from the last in the series; I wanted to keep the abort stuff last).  Starting my stress test locally to see if anything got missed or broken at the last minute, but I'm hopeful.

Commit:
kernel/include: Elaborate documentation for k_thread_{abort,join}()

Clarify some edge cases that struck me as important to understand.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
So, overnight stress on x86_64 showed that the msgq_api test is failing maybe 13% of the time in a loaded twister run.  This is a test that does a lot of racing aborts (it's one of the big offenders in the intermediate states in this PR), so it's likely it's showing some kind of remaining bug here.  I'll keep looking.  Other failures are happening randomly at the rate of one per ~20 test passes, and seem to be the kind of spurious thing we are used to on qemu due to timing interactions.  The msgq test seems to be the only one with a measurable failure rate on its own.

FWIW: that failure rate is still much, much better than what we have in tree.  So it's probably worth merging this anyway (and I guess opening a bug if so) pending whatever fix shows up for the msgq_api failures.

nashif commented on kernel/thread.c:
```diff
@@ -800,7 +800,7 @@ void z_init_thread_base(struct _thread_base *thread_base, int priority,
 		       uint32_t initial_state, unsigned int options)
 {
 	/* k_q_node is initialized upon first insertion in a list */
```
unrelated to the change, what is `k_q_node`? seems to be some old craft

andyross replied:
It's called "qnode_dlist" or "qnode_rb" now (depending on the scheduler queue backend selected).  This deserves a rework.  Really the whole "thread_base" struct could go away and be folded into thread, the only user of it is the weird dummy thread trick in pipe and mailbox, and they could have their own separate API what they want.

jharris-intel commented on tests/kernel/mem_protect/syscalls/prj.conf:
```diff
@@ -1,5 +1,5 @@
 CONFIG_TEST_USERSPACE=y
 CONFIG_TIMESLICING=y
-CONFIG_TIMESLICE_SIZE=10
+CONFIG_TIMESLICE_SIZE=20
```
Why the change? Working around lingering issues?

andyross replied:
This is the timeslicing fix.  There were tests that were asking for timeslices so small they couldn't be reliably scheduled by the timer driver, and a combination of recent changes to the timeslice layer (which for subtle reasons dropped integration some code that would elide the z_clock_set_timeout() call on imminent expirations) and more deterministic thread abort behavior meant that we were hitting the hang condition.

But really it's a bug in the tests.  Minimum timeslice size when CONFIG_TICKLESS_KERNEL is two ticks.  One won't work right (and now the kernel protects against it).

jharris-intel commented on tests/kernel/workq/critical/src/main.c:
```diff
@@ -166,7 +166,7 @@ void regression_thread(void *arg1, void *arg2, void *arg3)
 
 	TC_PRINT("Enable timeslicing at %u\n", k_uptime_get_32());
-	k_sched_time_slice_set(10, 10);
+	k_sched_time_slice_set(20, 10);
```
Minor: shouldn't this be pulled from .conf?

andyross replied:
Yes, but it wasn't before so it isn't now. :)  My personal feeling about having ugly code in the regression suite is that it's far more important that it stay there and run than experience churn that might hide bugs we'd otherwise detect.  Zephyr is expected to run bad code correctly too.

(Not like I'd reject a patch that cleaned up hardcoded timeslices or anything, just that it isn't my personal inclination to try to "clean up" tests at the same moment I'm trying to make them pass.)

andyross commented on arch/posix/core/posix_core.c:
```diff
@@ -507,16 +507,11 @@ void z_impl_k_thread_abort(k_tid_t thread)
-		z_self_abort();
-		CODE_UNREACHABLE; /* LCOV_EXCL_LINE */
+		z_thread_abort(_current);
+		return;
```
Actually yes, in interrupt context.  This is the native_posix implementation of k_thread_abort() and it will be called in all the same contexts the "official" one is.  But you're right, the ISR test is already covered above so it's silly to have the return here.

Commit:
drivers/timer/hpet: Extend qemu workaround

Qemu when running more than one processor has a known synchronization
bug where counter values read from the HPET (notionally a single
global device) can be seen going "backwards" when read from different
CPUs.

There was a pre-existing workaround in the ISR that knew about this,
but the problem can crop up anywhere the counter value is used.  In
particular I caught it aliasing with the "max_ticks" computation in
z_clock_set_timeout(), where it would cause a rollover and the
resulting negative comparator value would result in no end of
hilarity.

Wrap all access to the counter register with a counter() inline that
(when the workaround is enabled) forces the result to be monotonic by
clamping it to a minimum of one more than the previously read value.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Correctly clamp z_clock_set_timeout() argument

This function would correctly suppress attempts to set timeouts that
were too soon for the driver or farther out than what was already set,
but when it actually set the timeout it would use the requested value
and not clamp it to the minimum of it and the current timeout
expiration, leading to "too-long" timeouts being set at the driver.

In uniprocessor configurations, that turns out to have been benign
because something else would always come back along when timeout state
changed and fix the broken value before the expiration.

But in SMP, this opens up races.  For example, the idle thread on one
CPU can see that there are no active threads and schedule a maximum
value timeout at the same time as the other thread adds a new timeout
that expects a near-term expiration.  The broken code here would see
that the new timeout exists, decide that yes it needs to override, but
then set the K_TICKS_FOREVER value it got from the idle thread!

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
OK new version up.  This adds two new fixes to address what I was seeing with msgq_api (and again it turns out that neither were related to abort per se and were just pre-existing bugs tickled by different timing):

The HPET driver had a workaround in its ISR for a bug in qemu where the clock can go backwards, but there was another edge case where the same glitch could (very rarely, but I did see it happen!) cause a bad timeout to be set due to a rollover condition.

The main bug was a simple thinko in the timeout code about which value to set as the timeout.  This turns out (due to the structure of how things work) to have been hidden in uniprocessor setups, but on SMP it's a real race.  This was the source of almost all the instability (and it took FOREVER to find, ugh...)

And... this might be it folks.  My build box is through 8 iterated twister runs on x86_64 with no failures reported at all.  I'll leave it running as long as possible to see if I can catch something failing.

jharris-intel commented on include/kernel.h:
```diff
@@ -537,6 +539,17 @@ __syscall k_tid_t k_current_get(void);
+ * this is done via blocking the caller (in the same manner as
+ * k_thread_join()), but in interrupt context on SMP systems the
+ * implementation is required to spin for threads that are running on
+ * other CPUs.  Note that as specified, this means that on SMP
```
As a note for future development here, there is a way to deal with this, it's just not great.

(Not something to do in this PR though!)

(Namely, if you're in interrupt context and you are spinning waiting for a thread to be aborted, check to see if the thread you interrupted interrupted thread should be aborted, and if so handle it yourself. There are a bunch of ordering considerations, but that's the general gist of it.)

andyross replied:
Yeah, in the general case that becomes a full graph search for cycles .   Real devices (not ones running Zephyr, but then you never know...) exist with hundreds of CPUs.  You'd need to be willing to search that graph while holding a global scheduler lock literally in the middle of a context switch.  In an "RTOS". :)

FWIW: I went digging to see what Linux does here, and the answer is actually it doesn't.  There's no accessible API I can find to kill a process from an interrupt at all.  You kill processes by sending them signals so the abort is always done from their own kernel stack.  So even having documented the deadlock, I think we're ahead of the game here.

jharris-intel replied:
>Yeah, in the general case that becomes a full graph search for cycles .

Hmm. Perhaps we're thinking of something different here (or you've identified something I'm missing, which is entirely likely).

The check for "if the thread you interrupted should be aborted" is constant-time, assuming you check the thread's flag. No searching the graph with BigLock.

At a conceptual level:

1. Every time you go to abort a thread, you ensure that your target thread is marked as aborting before you drop into `join` (...which we already do).
2. If you're spinning in a join in interrupt context, and the interrupted thread is marked as aborting, handle that abort.

(The key here is that there _is_ an implicit graph search, but it's incremental and amortized across all of the `k_thread_abort` calls. See, the only way to _create_ the graph of aborting threads is with a call to `k_thread_abort` for each edge... and each node has flags. If you squint a lot this turns into graph labelling...)

(There are still problems here with e.g. aborting a thread that's holding a mutex I am well aware. But this at least solves the cycle-of-aborting-threads issue.)

andyross replied:
Ah, got it.  I was talking about being able to detect the cycle of parallel aborts and return an error to a user who tries.

But you're right, being able to detect a newly aborted _current from within that spin loop would have the effect of breaking the cycle.  That's clever, I'll see what needs to be done, but yeah, maybe in a separate PR.

jharris-intel commented on kernel/sched.c:
```diff
@@ -457,6 +447,25 @@ static void update_cache(int preempt_ok)
 
+static bool thread_active_elsewhere(struct k_thread *thread)
+{
+	/* True if the thread is currently running on another CPU.
```
I'm surprised threads don't have a "I'm running" flag.

(Then this could just be `thread_running(thread) && thread != _current`)

andyross replied:
That would be the obvious way to implement this, yeah.  But it requires allocating another flag bit, and we're out of space at the moment.  I'm sensitive to impacting the size of the thread struct for non-SMP use cases.  Really this isn't a hard problem to solve, but on existing devices a quick array loop over 2-4 CPUs is fine for now.

jharris-intel replied:
Hmm.

Looking at it, we could probably swap out a few of the boolean flags with a state enum instead - most of those flags are one-hot anyway.

But yes, this is fine for now I think.

****

This is _decidedly_ off-topic for this particular PR, but:

> Really this isn't a hard problem to solve, but on existing devices a quick array loop over 2-4 CPUs is fine for now.

My initial testing of Zephyr is showing the opposite, unfortunately. We're fairly bottlenecked on sched_lock in general. (E.g. 2 independent pairs of threads passing 2 semaphores back and forth on a 4-core system end up having _fewer_ total swaps per second than one pair of threads passing one semaphore back and forth.)

(I'm holding off on making a ticket for this until AArch64 SMP support is actually in trunk :-) . Correctness is a rather important prerequisite to performance.)

andyross replied:
I'd love to look at that report when you're ready.  There's surely stuff we can do to relax contention.  But do at least recognize that k_sem is a VERY thin layer, code that does nothing but pass semaphore tokens back and forth is in some sense doing nothing **but** scheduling, so absent a change in the fundamentals you'd expect the Zephyr design to produce a constant amount of scheduling regardless of CPU count.

jharris-intel replied:
Yep, it's a rather fundamental design decision that bottlenecks us here unfortunately :-(

In general, the "sea of spinlocks" approach - although better than BigLock - still results in individual features being sequentially-bottlenecked. And when one of those features is "scheduling"...

What it means in practice is that it's not much better than BigLock if you're doing anything that can schedule (and practically all interaction with Zephyr can schedule...)

You can do a whole lot better with "sea of spinlocks + atomics for base scheduling"... at the expense of a whole lot of complexity and nondeterminism. Most RTOSes I've seen don't even attempt this approach - which makes sense in a number of ways - but also means that they tend to be unsuitable for some usecases. The difference between 200 cycles to context-switch when you have a budget of 2000 cycles / core / op and 2000 cycles to context-switch when you have a budget of 2000 cycles / core / op is the difference between "you can carefully use the RTOS" and "you can't use the RTOS", in a lot of ways.

I may end up having to reject Zephyr here, which is rather unfortunate.

I can make a ticket for context-switch performance now to start the discussion, if you'd like. Just know that it's based on a hacked-together branch with MrVan's AArch64 SMP changes, your abort rewrite, some "trivial" local optimizations (notably: changing spinlock release to a sequentially-consistent store instead of a CAS non-trivially improves the performance of the spinlock in contended situations), and some proprietary driver code (nothing that should directly affect this, however)...

jharris-intel replied:
@andyross - I created #32686 with one of the contributing factors. The current spinlock has some rather terrible behavior under heavy contention (such as all cores hammering at sched_lock).

jharris-intel replied:
@andyross - I created #32708 with another one of the contributing factors. The current implementation of `k_current_get` is an unfortunate combination of "slow" and "called a lot".

jharris-intel reviewed (approved):
The parts that I followed make sense.

(I honestly didn't take the time to follow all of the logic here, so take this with a grain of salt.)

andyross commented on drivers/timer/hpet.c:
```diff
@@ -138,10 +148,14 @@ int z_clock_driver_init(const struct device *device)
 	max_ticks = (0x7fffffff - cyc_per_tick) / cyc_per_tick;
-	last_count = MAIN_COUNTER_REG;
+	last_count = counter();
+#ifdef GLITCHY_EMU
+	last_counter = last_count;
+	max_ticks -= 20;
+#endif
 
 	TIMER0_CONF_REG |= TCONF_INT_ENABLE;
-	TIMER0_COMPARATOR_REG = MAIN_COUNTER_REG + cyc_per_tick;
+	TIMER0_COMPARATOR_REG = counter() + cyc_per_tick;
```
It's an atomic register and this is the initialization path against which it won't be racing.  So yeah, this is safe.

jharris-intel replied:
Thanks for the doublecheck!

> It's an atomic register

It's more the (non-atomic) `GLITCHY_EMU` workaround I was wondering about.

andyross commented on drivers/timer/hpet.c:
```diff
@@ -41,18 +41,41 @@ DEVICE_MMIO_TOPLEVEL_STATIC(hpet_regs, DT_DRV_INST(0));
 
 #define MIN_DELAY 1000
 
+/* The clock on Qemu in SMP has been observed going backwards.  This
```
Without, we don't have it wired right now.  And yes, it's likely that this is specific to the realtime clock integration.  I never investigated, and actually had forgotten all about this workaround until I caught it misbehaving yesterday.

jharris-intel replied:
> Without

That'd explain it then I think... You might try setting `-rtc clock=vm`.

I think you might be running into ntp updating the host time out from under you.

andyross commented:
Came back to look at the stress log.  74 runs of "twister -p qemu_x86_64" with no failures.  I think this one is cooked for now.  Seeing rarer failures is going to need a more extensive rig.

lyakh commented on kernel/include/ksched.h:
```diff
@@ -67,6 +67,7 @@ void z_ready_thread(struct k_thread *thread);
 void z_thread_single_abort(struct k_thread *thread);
 FUNC_NORETURN void z_self_abort(void);
 void z_requeue_current(struct k_thread *curr);
+struct k_thread *z_swap_next_thread(void);
```
@andyross this commit broke the SMP sync example on UP2 - identified with bisect... So, currently it's broken on "master" But one of your patches from #32356 seems to fix this breakage again...
